docs/how-to-use-and-FAQ/use-ncnn-with-pytorch-or-onnx.md

```diff
@@ -2,8 +2,114 @@ Here is a practical guide for converting pytorch model to ncnn
+
+* C. build from source
+  1. install pytorch
+  2. (optional) install torchvision for pnnx torchvision operator support
+  3. (optional) install protobuf for pnnx onnx-zero support
+  4. clone https://github.com/Tencent/ncnn.git
+  5. build pnnx in ncnn/tools/pnnx with cmake
+
+  You will probably refer https://github.com/pnnx/pnnx/blob/main/.github/workflows/release.yml for detailed steps
+
+  ```shell
+  git clone https://github.com/Tencent/ncnn.git
+  mkdir ncnn/tools/pnnx/build
+  cd ncnn/tools/pnnx/build
+  cmake -DCMAKE_INSTALL_PREFIX=install -DTorch_INSTALL_DIR=<your libtorch install dir> -DTorchVision_INSTALL_DIR=<your torchvision install dir> ..
+  cmake --build . --config Release -j 4
+  cmake --build . --config Release --target install
+  ```
+
+### How to use pnnx?
+* A. python
+  1. optimize and export your torch model with pnnx.export()
+      ```python
+      import torch
+      import torchvision.models as models
+      import pnnx
+
+      model = models.resnet18(pretrained=True)
+
+      x = torch.rand(1, 3, 224, 224)
+
+      opt_model = pnnx.export(model, "resnet18.pt", x)
+
+      # use tuple for model with multiple inputs
+      # opt_model = pnnx.export(model, "resnet18.pt", (x, y, z))
+      ```
+  2. use optimized module just like the normal one
+      ```python
+      result = opt_model(x) 
+      ```
+  3. pick resnet18_pnnx.py for pnnx-optimized torch model
+  4. pick resnet18.ncnn.param and resnet18.ncnn.bin for ncnn inference
+
+B. command line
+  1. export your torch model to torchscript / onnx
+      ```python
+      import torch
+      import torchvision.models as models
+
+      net = models.resnet18(pretrained=True)
+      net = net.eval()
+
+      x = torch.rand(1, 3, 224, 224)
+
+      # You could try disabling checking when tracing raises error
+      # mod = torch.jit.trace(net, x, check_trace=False)
+      mod = torch.jit.trace(net, x)
+
+      mod.save("resnet18.pt")
+
+      # You could also try exporting to the good-old onnx
+      torch.onnx.export(net, x, 'resnet18.onnx')
+      ```
+
+  2. pnnx convert torchscript / onnx to optimized pnnx model and ncnn model files
+      ```shell
+      ./pnnx resnet18.pt inputshape=[1,3,224,224]
+      ./pnnx resnet18.onnx inputshape=[1,3,224,224]
+      ```
+      macOS zsh user may need double quotes to prevent ambiguity
+      ```shell
+      ./pnnx resnet18.pt "inputshape=[1,3,224,224]"
+      ```
+      For model with multiple inputs, use list
+      ```shell
+      ./pnnx resnet18.pt inputshape=[1,3,224,224],[1,32]
+      ```
+      For model with non-fp32 input data type, add type suffix
+      ```shell
+      ./pnnx resnet18.pt inputshape=[1,3,224,224]f32,[1,32]i64
+      ```
+  3. pick resnet18_pnnx.py for pnnx-optimized torch model
+  4. pick resnet18.ncnn.param and resnet18.ncnn.bin for ncnn inference
+
+see more pnnx informations: https://github.com/pnnx/pnnx
 
+## pytorch to onnx
```

Comment:
```suggestion
## pytorch to onnx (deprecated)
```

Reply:
this suggestion was successfully applied.